### ISSUE TITLE
docker push uses too much of my bandwidth

### ISSUE THREAD

**Issue by mcandre** [docker/docker#13342]
My 60mbps home Internet connection can't handle docker pushing very well.

* docker push often freezes at 50-75% for an hour or more, sometimes failing.
* docker push always hogs all my bandwidth until it exits. My roommates can't play online video games, and I can't even browse docker docs to diagnose the problem until my docker push finishes.

System:

```
$ specs docker os
specs --version
0.19

docker version
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): darwin/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64

boot2docker version
Boot2Docker-cli version: v1.6.0
Git commit: 9894ae9

vagrant --version
Vagrant 1.7.2

vboxwebsrv --help 2>&1 | grep VirtualBox
Oracle VM VirtualBox web service Version 4.3.24

system_profiler SPSoftwareDataType | grep 'System Version'
    System Software Overview:
      System Version: OS X 10.10.3 (14D136)
```

**Comment by jmakanjuola** [comment]
mcandre,

What's your router brand? I think it has sometime to do with your network. When was the last time you rebooted your router and modem? Reboot them and see if it makes a difference.

**Comment by thaJeztah** [comment]
If not a network issue, something like a `--throttle` flag to limit network traffic would be what you're looking for?

**Comment by mcandre** [comment]
Router: Belkin AC750 with 801.11ac.

Yes, a `--throttle` option would be much appreciated!

If it works well, I'd like to see it become the default eventually.

Here's the rest of the information:

```
$ docker version
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): darwin/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64

$ docker info
Containers: 1
Images: 182
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 184
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 3.18.11-tinycore64
Operating System: Boot2Docker 1.6.0 (TCL 5.4); master : a270c71 - Thu Apr 16 19:50:36 UTC 2015
CPUs: 4
Total Memory: 1.961 GiB
Name: boot2docker
ID: MCKR:WP3W:YXR7:M66R:E4QC:CSQZ:ZQAJ:E6D7:TXNH:P7J5:2RAT:TB4K
Debug mode (server): true
Debug mode (client): false
Fds: 19
Goroutines: 32
System Time: Wed May 20 22:35:38 UTC 2015
EventsListeners: 0
Init SHA1: 9145575052383dbf64cede3bac278606472e027c
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Username: mcandre
Registry: [https://index.docker.io/v1/]

$ uname -a
Darwin mockingjay.local 14.3.0 Darwin Kernel Version 14.3.0: Mon Mar 23 11:59:05 PDT 2015; root:xnu-2782.20.48~5/RELEASE_X86_64 x86_64
```

**Comment by thaJeztah** [comment]
@mcandre I had a quick chat with @crosbymichael about this issue on IRC and it's most likely something related to your network (setup). Docker causing the whole network to lock up is not something that should happen in normal situations.

I also polled for interest in adding a `--throttle` or `--cap` feature, but that's not something that's likely to appear in docker.

I am going to close this issue, because this is unlikely to be a bug in Docker, but probably caused by something in your network.

Feel free to continue this discussion here after I closed this. I hope you are able to find the cause.

**Comment by crosbymichael** [comment]
Just as some extra help, your router probably have QoS controls for limiting traffic to certain ports,apps, etc.  This is usually the best way to handle this on your type of network.  Most configuration for things like this either happen at the hardware level or at the system level.  It's not managed at the application level because it would be duplicated effort.  Torrenting, uploading/watching youtube, etc, all these applications would have the same issues if you have a network problem.  

I'd looking into your router's QoS settings first.

**Comment by mcandre** [comment]
Yes, enabling my router's QoS settings helps prevent docker from stealing too much of my bandwidth.

**Comment by codekitchen** [comment]
I'm not sure it'll really change the decision to do nothing about this, but as another data point, this appears to be a common issue on many public/hotel networks where changing the QoS settings isn't feasible. Our remote, distributed development team has been complaining about docker quite loudly due to this.

**Comment by mcandre** [comment]
What if we defaulted docker push/pull bandwidth to a throttled setting (I'm sure DockerHub and CoreOS would appreciate it!), and gave users the option to override this setting in a per-user config file?

**Comment by jamiekurtz** [comment]
I'm having similar issues when pushing to a private repo on Docker Hub from an EC2 Ubuntu server. After a few minutes, I can no longer connect to my box (via SSH or Jenkins web site running on it), and have to stop and restart the server from the AWS console.

Is there any way to reduce the network demand when doing docker pushes/pulls? At home,  I had to cap outbound traffic on my pfsense firewall to keep `docker push` from killing my Brighthouse broadband connection. And now on EC2 a simple push is making the box completely inaccessible.

**Comment by dmitrym0** [comment]
[trickle](https://wiki.archlinux.org/index.php/Trickle) seems like a potential solution:

> a portable lightweight userspace bandwidth shaper

**Comment by dts** [comment]
This is a bit ridiculous, IMO.  Docker now defaults to pushing as many layers simultaneously as possible.  This is bad - layers can be 100s of megs, why would the default be to push many at the same time, retrying after 30s if one fails?  I can no longer push anything from my home computer, even with my 6MB/s up connection.  I would argue that this is not "the network's fault", or some QoS setting on the router.  What is the reasoning behind simultaneous uploading, other than "saturate the uplink" - because that is exactly what it does, at the cost of all other things on the network (including, say, web browsing on the client's computer).

**Comment by aaronlehmann** [comment]
> This is a bit ridiculous, IMO. Docker now defaults to pushing as many layers simultaneously as possible.

There is a limit of three simultaneous uploads per push.

> This is bad - layers can be 100s of megs, why would the default be to push many at the same time, retrying after 30s if one fails?

Most layers are quite small. However, a layer upload can still take a significant amount of time, because there are multiple round-trips involved, and registry storage backends can have high latency. Handling more than one layer in parallel helps avoid dead time between layers, and in practice makes pushes substantially faster.

> I can no longer push anything from my home computer, even with my 6MB/s up connection. I would argue that this is not "the network's fault", or some QoS setting on the router. What is the reasoning behind simultaneous uploading, other than "saturate the uplink" - because that is exactly what it does, at the cost of all other things on the network (including, say, web browsing on the client's computer).

This is surprising. The only thing that would cause a push to time out is if a connection for one of the layers can't be opened, even after repeated retries. It sounds like when your connection is saturated, it's impossible to do DNS lookups and/or open TCP connections. When you upload a file using a different tool like scp, are you able to browse the web?

**Comment by stevvooe** [comment]
@dts This issue has been closed since May of last year. I'd really recommend opening a separate issue describing your particular problem. 

Typically, the kinds of problems you are describing are best handled with proper QoS configuration. It sounds like there may be some unfortunate timeouts that don't work well in your network configuration. Let's have you open an issue and we'll see if we can track this down.

**Comment by dts** [comment]
@stevvoe - I've filed a separate ticket here (https://github.com/docker/docker/issues/20936).

@aaronlehmann - Most of the images I work with have at least 3 or 4 100+MB layers.  Everything except these make it through no problem, but then they all sit there fighting for bandwidth, and each of them times out while the others push, and on and on, until everything fails.  Every time they time out they retry from the beginning, so this cycle almost never succeeds completely.

**Comment by sehugg** [comment]
Hey, @jamiekurtz ... I have Brighthouse too! And I have the exact same problem. For now, I'm using wondershaper to work around the inadequate upload speeds and the plausibly-deniable traffic shaping.

**Comment by thaJeztah** [comment]
For those following this issue; there's a pull request that allows limiting the number of simultaneous uploads/downloads per pull/push https://github.com/docker/docker/pull/22445, which will be part of docker 1.12. Although this is not a real "throttling" feature, it may help the issue that's discussed here

**Comment by dmitrym0** [comment]
[Here's a description of how to throttle arbitrary processes Mac OS X](https://dreness.com/blog/archives/843).

**Comment by jcelerier** [comment]
I have the same issue as OP : doing `docker push` just bricks my router's internet access part (french ISP router), and I have to reboot it to get it to work again. I tried running with `trickle` as some mentioned : 

    # trickled -f -d 1000 -u 1000 
    # docker push ...

but it does not seem to be affected.

**Comment by Guidonet** [comment]
Also noticed the same issue today as well.  I get i can do QOS, but i don't get why this is the only app in recent memory that causes something like this.

**Comment by zoobab** [comment]
The reason why trickle does not do anything it is because it is intercepting gLIBC syscalls, and docker is written in GO, so it has no impact whatsoever.

What worked somehow for me was a traffic shapping script, where you have to play with some values, and verify the speed with iptraf-ng to see what kind of traffic shapping you can get:

http://serverfault.com/questions/191560/how-can-i-do-traffic-shaping-in-linux-by-ip

**Comment by jcelerier** [comment]
I also tried with `tc` and it did not work either

**Comment by ojamui** [comment]
Does anyone has a solution? ive yet to find any method to limit the docker pull download rate

**Comment by thaJeztah** [comment]
@ojamui https://github.com/docker/docker/issues/13342#issuecomment-217671222 may help

**Comment by EronWright** [comment]
TLDR; update your docker daemon configuration to restrict the # of concurrent uploads:
```
"max-concurrent-uploads" : 1
```

**Comment by RPGillespie6** [comment]
I ran into this problem as well. Pushing from my home PC completely froze the entire network (kids and wife complaining that the internet stopped working). On top of that the push would always fail. @EronWright 's solution worked perfectly. Simply add the following to `/etc/docker/daemon.json`:
`"max-concurrent-uploads" : 1`

Should look something like:

```
{
    "max-concurrent-uploads" : 1
}
```

**Comment by Ark74** [comment]
Thanks @EronWright and @RPGillespie6 even when this won't limit the bandwidth use (it will keep eating all the bandwidth available) it prevents the pushing from failing.

Thanks you guys!

**Comment by markkolich** [comment]
In case it wasn't obvious to the reader (it wasn't to me):

* On macOS, the relevant `daemon.json` file can be found at `~/.docker/daemon.json` .... you'll want to add `"max-concurrent-uploads" : 1` to the existing JSON object therein.
* You need to restart the Docker daemon for this change to take effect ... sans some spiffy command line sequence, on Mac that seems to involve just quitting Docker then reopening it.

**Comment by thaJeztah** [comment]
If you're preferring a "UI"; on docker for mac / docker for windows there's also a tab in the configuration panel to edit the file; https://docs.docker.com/docker-for-mac/#daemon-configuration-file

![img_0771](https://user-images.githubusercontent.com/1804568/37565347-168e567a-2aa8-11e8-983c-fabd9de7bc36.PNG)

**Comment by mscottx88** [comment]
QoS changes in the router worked great for me.   I've been seeing failures over time, forcing me to reboot my modem/router.  Now, I'm good.

**Comment by UmutAlihan** [comment]
> This is a bit ridiculous, IMO. Docker now defaults to pushing as many layers simultaneously as possible. This is bad - layers can be 100s of megs, why would the default be to push many at the same time, retrying after 30s if one fails? I can no longer push anything from my home computer, even with my 6MB/s up connection. I would argue that this is not "the network's fault", or some QoS setting on the router. What is the reasoning behind simultaneous uploading, other than "saturate the uplink" - because that is exactly what it does, at the cost of all other things on the network (including, say, web browsing on the client's computer).

Aggree 100%. Whenever I try to push or pull, I totally lost control over that server, since I only can connect headless via SSH, I require that small bandwith to send commands via secure shell.

But pulling or pushing layers simultaneously somehow blocks all the bandwith, which I don't encounter this issue via another  uploading/downloading scenarios.

This is most probably a wrong or narrow thought architectural decision or it is also another possibility that docker does not about hobbyists who utilize consumer-grade low-level hardware and connections, only enterprise-grade high-end connected large corporations...
